### PR TITLE
linux imgui

### PR DESCRIPTION
**Pull request by xezon**

### CODE REVIEW

File: BUILDING.md [hunk 1]
```diff
@@ -1,67 +1,126 @@
...
-                  hexidecimal notation.
-  -e --end        Ending address of a single function to dissassemble in
-                  hexidecimal notation.
-  -v --verbose    Verbose output on current state of the program.
-  --section       Section to target for dissassembly, defaults to '.text'.
-  --listsections  Prints a list of sections in the exe then exits.
-  -d --dumpsyms   Dumps symbols stored in the executable to the config file.
-                  then exits.
-  -h --help       Displays this help.
+  unassemblize [OPTION...] positional parameters
```
Comment (xezon) [inline BUILDING.md]: This command line text runs risk of being outdated. Perhaps we just mention to view the help text?

File: BUILDING.md [hunk 2]
```diff
@@ -1,67 +1,126 @@
...
+docker run -v $(pwd):/work unassemblize [OPTIONS] [INPUT]
 ```
 
-Run executable:
+## Usage
+
 ```
 $ ./unassemblize
 
 unassemblize r6 ~201fc23
```
Comment (xezon) [inline BUILDING.md]: Revision is outdated. Perhaps do not mention it.

File: CMakeLists.txt [hunk 3]
```diff
@@ -54,6 +63,20 @@ set(LIEF_ART OFF CACHE BOOL "" FORCE)
...
+    GIT_TAG 3.3.9
+)
+
+# Configure GLFW options
+set(GLFW_BUILD_DOCS OFF CACHE BOOL "" FORCE)
+set(GLFW_BUILD_TESTS OFF CACHE BOOL "" FORCE)
+set(GLFW_BUILD_EXAMPLES OFF CACHE BOOL "" FORCE)
+set(GLFW_INSTALL OFF CACHE BOOL "" FORCE)
+
+FetchContent_MakeAvailable(glfw)
```
Comment (xezon) [inline CMakeLists.txt]: Is this fetched on Windows? If yes, is that necessary?

File: CMakeLists.txt [hunk 4]
```diff
@@ -162,16 +185,28 @@ target_sources(unassemblize PRIVATE
...
+    )
+else()
+    target_sources(unassemblize PRIVATE
+        src/imguiclient/imguiglfw.cpp
+        src/imguiclient/imguiglfw.h
+    )
+endif()
+
 target_link_libraries(unassemblize PRIVATE
+    imgui_unix
```
Comment (xezon) [inline CMakeLists.txt]: Should this be excluded from Windows?

File: src/commontypes.h [hunk 5]
```diff
@@ -14,6 +14,7 @@
 
 #include <array>
 #include <assert.h>
+#include <cstring>
```
Comment (xezon) [inline src/commontypes.h]: What does this do?

File: src/imguiclient/imguiapp.h [hunk 6]
```diff
@@ -13,11 +13,13 @@
 #pragma once
 
 // clang-format off
-#include <imgui_internal.h>
 #include <imgui.h>
+#ifdef _WIN32
+#include <imgui_internal.h>
```
Comment (xezon) [inline src/imguiclient/imguiapp.h]: Looks like we do not need this anymore. I tested on Windows.

File: src/imguiclient/imguiglfw.cpp [hunk 7]
```diff
@@ -0,0 +1,346 @@
...
+    return true;
+}
+
+void ImGuiGLFW::shutdown()
+{
+    if (!m_initialized)
+        return;
+
+    // Cleanup ImGui app
+    if (m_app)
```
Comment (xezon) [inline src/imguiclient/imguiglfw.cpp]: No need to test.

File: src/main.cpp [hunk 8]
```diff
@@ -329,16 +331,16 @@ int main(int argc, char **argv)
...
-
     if (g_options.gui)
     {
 #ifdef WIN32
         unassemblize::gui::ImGuiWin32 gui;
         unassemblize::gui::ImGuiStatus status = gui.run(g_options);
         return int(status);
 #else
-        gui_error = true;
+        unassemblize::gui::ImGuiGLFW gui;
```
Comment (xezon) [inline src/main.cpp:343]: This works on all platforms?

File: CMakeLists.txt [hunk 9]
```diff
@@ -180,30 +215,33 @@ target_link_libraries(unassemblize PRIVATE
...
-)
-
 if(WINDOWS)
-    target_sources(unassemblize PRIVATE
-        wincompat/strings.h
-    )
-
     target_link_libraries(unassemblize PRIVATE
-        d3d9.lib # How does it find it???
-        imgui_win32
```
Comment (xezon) [inline CMakeLists.txt:196]: Where is this linked now?

Comment (bobtista) [reply CMakeLists.txt:196]: Mistakenly removed - it's now been restored

File: src/main.cpp (hunk 8 above)
Comment (bobtista) [reply src/main.cpp:343]: OpenGL + GLFW in theory works on windows, mac, and linux. I haven't tested in on my windows machine yet though. Also, I do like having an else gui_error = true block as the default if someone is using something other than windows mac or linux - so I added more specific ifs, and restored the else error handling from before.

File: src/imguiclient/imguiglfw.cpp (hunk 7 above)
Comment (bobtista) [reply src/imguiclient/imguiglfw.cpp]: Good catch. Removed.

File: src/imguiclient/imguiapp.h (hunk 6 above)
Comment (bobtista) [reply src/imguiclient/imguiapp.h]: Ok I've removed it.

File: src/commontypes.h (hunk 5 above)
Comment (bobtista) [reply src/commontypes.h]: False positive - I thought this change was fixing a build error on Ubuntu, but it still builds after removing it. It's now removed

File: CMakeLists.txt (hunk 4 above)
Comment (bobtista) [reply CMakeLists.txt]: Yes, I updated cmakelists to keep the non-windows items separate .

File: CMakeLists.txt (hunk 3 above)
Comment (bobtista) [reply CMakeLists.txt]: No need, I've updated this so it doesn't happen on windows

File: BUILDING.md (hunk 2 above)
Comment (bobtista) [reply BUILDING.md]: Removed

File: BUILDING.md (hunk 1 above)
Comment (bobtista) [reply BUILDING.md]: Done

File: CMakeLists.txt [hunk 10]
```diff
@@ -54,6 +63,22 @@ set(LIEF_ART OFF CACHE BOOL "" FORCE)
...
 set(LIEF_VDEX OFF CACHE BOOL "" FORCE)
 
+# Fetch GLFW 3.3
+FetchContent_Declare(glfw
+    GIT_REPOSITORY https://github.com/glfw/glfw.git
+    GIT_TAG 3.3.9
+)
+
+# Configure GLFW options
+if(NOT WIN32)
```
Comment (xezon) [inline CMakeLists.txt]: Some places use WINDOWS, others WIN32. Can you streamline this?

File: CMakeLists.txt [hunk 11]
```diff
@@ -162,16 +187,50 @@ target_sources(unassemblize PRIVATE
...
+else()
+    target_sources(unassemblize PRIVATE
+        src/imguiclient/imguiglfw.cpp
+        src/imguiclient/imguiglfw.h
+    )
+endif()
+
+if(WINDOWS)
+    target_link_libraries(unassemblize PRIVATE
+        d3d9.lib
```
Comment (xezon) [inline CMakeLists.txt]: This was commented before with "# How does it find it???". Perhaps we should retain a TODO comment that says to figure out if any "FindDirectX" logic is needed here.

File: src/functiontypes.cpp [hunk 12]
```diff
@@ -13,6 +13,7 @@
 #include "functiontypes.h"
 #include "util.h"
 #include <assert.h>
+#include <cstring>
```
Comment (xezon) [inline src/functiontypes.cpp]: What does this do?

File: src/imguiclient/imguiapp.h [hunk 13]
```diff
@@ -13,11 +13,10 @@
 #pragma once
 
 // clang-format off
-#include <imgui_internal.h>
 #include <imgui.h>
 // clang-format on
```
Comment (xezon) [inline src/imguiclient/imguiapp.h]: Can now remove the clang-format off on comments.

File: src/util.h [hunk 14]
```diff
@@ -15,6 +15,8 @@
 #include <string>
 #include <string_view>
 #include <vector>
+#include <algorithm>
+#include <cstdint>
```
Comment (xezon) [inline src/util.h]: Was this necessary?

File: CMakeLists.txt [hunk 15]
```diff
@@ -40,6 +40,15 @@ find_package(ClangFormat)
 include(ClangFormat)
 include(FetchContent)
 
+if(NOT WINDOWS)
```
Comment (xezon) [inline CMakeLists.txt]: Perhaps can find a better platform name than `NOT WINDOWS`, for example `APPLE AND LINUX` if that works.

File: CMakeLists.txt (hunk 10 above)
Comment (bobtista) [reply CMakeLists.txt]: Done, also organized the WINDOWS vs APPLE OR UNIX stuff better

File: CMakeLists.txt (hunk 11 above)
Comment (bobtista) [reply CMakeLists.txt]: Done

File: src/functiontypes.cpp (hunk 12 above)
Comment (bobtista) [reply src/functiontypes.cpp]: There were a few things that I thought were fixing a build error on Ubuntu, but I suspect something else was the issue. I've removed the new includes like this one.

File: src/imguiclient/imguiapp.h (hunk 13 above)
Comment (bobtista) [reply src/imguiclient/imguiapp.h]: Done

File: src/util.h (hunk 14 above)
Comment (bobtista) [reply src/util.h]: Removed

File: CMakeLists.txt (hunk 15 above)
Comment (bobtista) [reply CMakeLists.txt]: I've changed it to APPLE OR UNIX, as UNIX is a cmake variable that includes Linux and Apple. To that end, we could just use if(UNIX) if we want

**Comment by xezon** [comment]
Can you please allow me to push to this branch? There were Windows fixes necessary.

I suspect I cannot push because I opened this Pull Request. You need to click a button somewhere that says that others can push to your branch.

**Comment by bobtista** [comment]
> Can you please allow me to push to this branch? There were Windows fixes necessary.
> 
> I suspect I cannot push because I opened this Pull Request. You need to click a button somewhere that says that others can push to your branch.

I've added you as a collaborator

**Pull request merged**